File: Data Structures & Algorithms/balanced-binary-tree/submission-3.go
func isBalanced(root *TreeNode) bool {
	return height(root) != -1
}

func height(root *TreeNode) int {
    if root == nil {
        return 0
    }
    left := height(root.Left)
    if left == -1 {
        return left
    }
    right := height(root.Right)
    if right ==-1 {
        return right
    }
    if abs(right-left) > 1 {
        return -1
    }
    return max(right, left)+1
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}

File: Data Structures & Algorithms/house-robber-iii/submission-3.py
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        cache = {}
        def dfs(root, taken):
            if not root:
                return 0
            if (root, taken) in cache:
                return cache[(root, taken)]
            res = 0
            if taken:
                res = dfs(root.left, False) + dfs(root.right, False)
            else:
                skip = dfs(root.left, False) + dfs(root.right, False)
                add = root.val + dfs(root.left, True) + dfs(root.right, True)
                res = max(skip, add)
            cache[(root,taken)] = res
            return res
            
        return dfs(root, False)